### PR TITLE
[move-prover] add strong edges for vector writebacks

### PR DESCRIPTION
**Pull request by JakeSilverman**
…

## Motivation

Strong edges for vector writebacks should further improve verification runtime. This commit also fixes a bug in the previous version of strong direct reference writebacks.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Plan to add some unit tests in to the bytecode test suite in future commit. This was tested on all existing benchmarks. Also recall that strong writebacks are off by default when using the move-prover, so this won't affect any integration testing at this time.

### CODE REVIEW

**Comment by JakeSilverman** [comment]
Note that I'm not really in love with the current abstract domain for edges (EdgeDomain in borrow_analysis). None of the three three strong edge kinds will ever belong to the same edge set, so the domain is perhaps a bit counterintuitive. But I believe implementation is cleaner by having the domain as currently is.

File: language/move-prover/bytecode/src/borrow_analysis.rs [hunk 1]
```diff
@@ -420,6 +421,35 @@ impl<'a> BorrowAnalysis<'a> {
             })
             .collect();
     }
+
+    fn get_vec_bor_mut_mid(&self) -> Option<move_model::model::ModuleId> {
```
Comment (wrwg) [inline language/move-prover/bytecode/src/borrow_analysis.rs:425]: Given that this linear search will happen with every single function call we visit, we can't do it like this.

Please precompute the `QualifiedId<FunId>` of the borrow_vec_mut function and store it in the state of the borrow analysis. Then the search would still happen for each function we analysis, so better even, store it as an extension in the global env (GlobalEnv::set_extension). You can do that by adding fn initialize to the FunctionTargetProcessor trait implementation.

**Review by wrwg — CHANGES_REQUESTED** [review]

File: language/move-prover/bytecode/src/borrow_analysis.rs (hunk 1 above)
Comment (JakeSilverman) [reply language/move-prover/bytecode/src/borrow_analysis.rs:425]: Nice catch. Meant to memoize this but completely forgot. Adding to globalenv sounds even better. Will do that.

File: language/move-prover/bytecode/src/borrow_analysis.rs [hunk 2]
```diff
@@ -342,6 +347,27 @@ impl FunctionTargetProcessor for BorrowAnalysisProcessor {
         data
     }
 
+    fn initialize(&self, _env: &GlobalEnv, _targets: &mut FunctionTargetsHolder) {
```
Comment (wrwg) [inline language/move-prover/bytecode/src/borrow_analysis.rs:350]: Remove `_`

**Review by wrwg — APPROVED** [review]

**Comment by JakeSilverman** [comment]
\land

**Pull request merged**